### PR TITLE
Doc: bump to Python 3.10; Sphinx 5.3.0. Remove `--constraint: foo -none` & some polishing

### PR DESCRIPTION
**Pull request by andreasabel**
User's guide: remove `-none` from constraint section, make `installed` more prominent.

- `PACKAGE -none` as already been removed in 3.4 as valid constraint. (I guess this is now flag syntax.)

- Important constraint `installed` was just buried in an example, I made it stand out more in a bullet point.
  Same for `source`, even though this is an exotic constraint.

- Highlight constraint `installed` in connection with `allow-newer`.

- Group `test` and `bench` constraint with the `+/-FLAG` constraints.

Rendering here: https://cabal--8789.org.readthedocs.build/en/8789/setup-commands.html#cmdoption-runhaskell-Setup.hs-configure-constraint

Also: bump to Python 3.10 for building the user's guide.

### CODE REVIEW

**Comment by Mikolaj** [comment]
A candidate for backport to 3.10?

File: doc/setup-commands.rst [hunk 1]
```diff
@@ -924,18 +916,30 @@ Miscellaneous options
...
     By default, constraints only apply to build dependencies
     (:pkg-field:`build-depends`), build dependencies of build
     dependencies, and so on. Constraints normally do not apply to
     dependencies of the ``Setup.hs`` script of any package
     (:pkg-field:`custom-setup:setup-depends`) nor do they apply to build tools
     (:pkg-field:`build-tool-depends`) or the dependencies of build
     tools. To explicitly apply a constraint to a setup or build
-    tool dependency, you can add a qualifier to the constraint as
-    follows:
+    tool dependency, you can add a qualifier ``setup`` or ``any``
```
Comment (ulysses4ever) [inline doc/setup-commands.rst:941]: I really dislike this style when, in this case, the definitions of `any` and `setup` are given in the comments of an example. There's description in the text above but it doesn't name the options!

File: doc/setup-commands.rst [hunk 2]
```diff
@@ -880,37 +880,29 @@ Miscellaneous options
...
 
-    To specify multiple constraints, you may pass the
-    ``constraint`` option multiple times.
+      - ``installed`` to fix a package to the already installed version.
+        Often useful for GHC-supplied packages in combination with :cfg-field:`allow-newer`,
+        e.g., ``--allow-newer='*:base' --constraint='base installed'``.
 
-    There are also some more specialized constraints, which most people
-    don't generally need:
+      - ``source`` to a fix a package to the local source copy.
```
Comment (ulysses4ever) [inline doc/setup-commands.rst:905]:
```suggestion
      - ``source`` to fix a package to the local source copy.
```

**Review by ulysses4ever — APPROVED** [review]

**Comment by ulysses4ever** [comment]
> A candidate for backport to 3.10?

Yes, please!

Thanks a lot for taking the time!

File: doc/setup-commands.rst (hunk 1 above)
Comment (andreasabel) [reply doc/setup-commands.rst:941]: Mmh, I don't like syntax _just_ be introduced by examples.  I like the style _overview followed by details_.  Graphically, I want the syntax to stick out, and in the example, everything is highlighted the same, so the emphasis on the important new syntax isn't there.

Comment (andreasabel) [reply doc/setup-commands.rst:941]: There _is_ highlighting of `any` and `setup` in the example due to putting them in single quotes and using the `console` highlighter.  However, this highlighter is broken in that it puts __strong__ emphasis on irrelevant words like _if_ and _in_ in the highlighted comment.  I find this very irritating if my attention is strongly drawn to purely administrative words in a sentence.  
A bigger rewrite of this section would move out the text from the console comments into the main body, fixing both your and my concern.  However, I don't want to go so far now.

**Comment by Mikolaj** [comment]
@mergify backport 3.10

File: doc/setup-commands.rst (hunk 1 above)
Comment (ulysses4ever) [reply doc/setup-commands.rst:941]: I just think that it's a bit of a pity to have a very nice and short explanation in comments as opposed to the main text. But I don't have any concrete suggestion here, so perhaps it's all right.

**Comment by andreasabel** [comment]
I'll update this to `master` and also try to get the bump in that is contained in this automatic PR:
- #8807

**Comment by andreasabel** [comment]
I forced `certifi >= 2022.12.7` in the `requirements.in` file now subsume #8807.

**Review by Mikolaj — APPROVED** [review]
Thank you.

**Comment by andreasabel** [comment]
Sphinx 6 is too new for readthedocs.org, but they support 5.3.0.

**Comment by Mikolaj** [comment]
Let me fast-track merge to make the 3.10 pre-release ready ASAP.

**Comment by Mikolaj** [comment]
@mergify rebase

**Pull request merged**